gordon_janitor/plugins_loader.py

```diff
@@ -0,0 +1,206 @@
+    # drill down to get config that matches plugin namespace
+    while plugin_config_keys:
+        key = plugin_config_keys.pop(0)
+        config = config.get(key)
+
+    # find which config namespaces map to other plugins
+    plugin_config_keys_to_exclude = set()
+    for plugin in all_plugins:
+        if plugin == plugin_namespace:
+            continue
+        if plugin.startswith(plugin_namespace):
+            # exclude same level & lower config keys for other plugins
+            keys_to_exclude = plugin.lstrip(plugin_namespace).split('.')
+            plugin_config_keys_to_exclude.update(keys_to_exclude)
+
+    # clean up config by removing other plugin configs
+    plugin_config = config.copy()
+    while plugin_config_keys_to_exclude:
+        key = plugin_config_keys_to_exclude.pop()
+        try:
+            plugin_config.pop(key)
+        except KeyError:
+            pass
+
+    return plugin_namespace, plugin_config
+
+
+def _load_plugin_configs(plugin_names, config):
+    # A plugin should only have access to its own config and
+    # parent/global plugin config, but not configs of other plugins
+    plugin_configs = {}
+    for plugin in plugin_names:
+        namespace, plugin_conf = _get_namespaced_config(
+            config, plugin, plugin_names)
+        plugin_configs[namespace] = plugin_conf
+
+    # merge namespaced with parent / global plugin config
+    merged_configs = {}
+    for namespace in plugin_configs.keys():
+        plugin_config = _merge_config(plugin_configs, namespace)
+        merged_configs[namespace] = plugin_config
+
+    return merged_configs
+
+
+def _get_plugin_config_keys(plugins):
+    # Make sure all parent namespaces of a plugin are available
+    # to load config for easy config handling
+    defined_plugin_namespace = plugins.keys()
+    all_config_keys = set()
+    for namespace in defined_plugin_namespace:
+        namespaces = namespace.split('.')
+        namespaces_to_build = []
+        while len(namespaces):
+            namespace = namespaces.pop(0)
+            namespaces_to_build.append(namespace)
+            config_key = '.'.join(namespaces_to_build)
+            all_config_keys.add(config_key)
+    return sorted(list(all_config_keys))
+
+
+def _get_activated_plugins(config, installed_plugins):
+    activated_plugins = config.get('core', {}).get('plugins', [])
+    for active_plugin in activated_plugins:
+        if active_plugin not in installed_plugins:
+            msg = f'Plugin "{active_plugin}" not installed'
+            raise exceptions.LoadPluginError(msg)
+    return activated_plugins
+
+
+def _gather_installed_plugins():
+    gathered_plugins = {}
+    for entry_point in pkg_resources.iter_entry_points(PLUGIN_NAMESPACE):
+        gathered_plugins[entry_point.name] = entry_point
+    return gathered_plugins
+
+
+def load_plugins(config):
+    """
+    Discover and instantiate plugins.
+
+    Args:
+        config (dict): loaded configuration for the Gordon service.
```

Comment:
Another word nitpick: `s/Gordon/Gordon-Janitor/` here and in `test_main.py` on line 119 and possibly on line 197 in the Returns part of this docstring ('_activated in gordon config._' -> '_activated in gordon-janitor config._`.